enterprise/server/scheduling/task_router/task_router.go

```diff
@@ -128,61 +136,118 @@ func (tr *taskRouter) RankNodes(ctx context.Context, cmd *repb.Command, remoteIn
 	if _, err := pipe.Exec(ctx); err != nil {
 		log.Errorf("Failed to mark task complete: redis pipeline failed: %s", err)
 		return
 	}
 
-	log.Debugf("Preferred executor %q added to %q", executorID, key)
+	log.Debugf("Preferred executor %q added to %q", executorID, routingKey)
+}
+
+// Contains the parameters required to make a routing decision.
+type routingParams struct {
+	cmd                *repb.Command
+	remoteInstanceName string
+	groupID            string
+}
+
+func getRoutingParams(ctx context.Context, env environment.Env, cmd *repb.Command, remoteInstanceName string) routingParams {
+	groupID := interfaces.AuthAnonymousUser
+	if u, err := perms.AuthenticatedUser(ctx, env); err == nil {
+		groupID = u.GetGroupID()
+	}
+	return routingParams{cmd: cmd, remoteInstanceName: remoteInstanceName, groupID: groupID}
 }
 
-// getPreferredNodeLimit returns the max number of nodes that should be stored
-// in the preferred executors list for each task key, as well as the max number
-// of preferred nodes that should be returned by RankNodes.
-func getPreferredNodeLimit(cmd *repb.Command) int {
-	isRunnerRecyclingEnabled := platform.IsTrue(platform.FindValue(cmd.GetPlatform(), platform.RecycleRunnerPropertyName))
-	if !isRunnerRecyclingEnabled {
-		return 0
+// Selects and returns a routingStrategy to use, or nil if none applies.
+func (tr taskRouter) selectRoutingStrategy(params routingParams) routingStrategy {
+	for _, strategy := range tr.strategies {
+		if strategy.applies(params) {
+			return strategy
+		}
 	}
-	workflowID := platform.FindValue(cmd.GetPlatform(), platform.WorkflowIDPropertyName)
+	return nil
+}
+
+func copyNodes(nodes []interfaces.ExecutionNode) []interfaces.ExecutionNode {
+	out := make([]interfaces.ExecutionNode, len(nodes))
+	copy(out, nodes)
+	return out
+}
+
+func minInt(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// A routingStrategy encapsulates a strategy for performing task routing based
+// on a set of provided routingParams. Currently this interface is somewhat
+// coupled with the Redis reading and writing logic in the task_router, but
+// that could be uncoupled in the future when other routing strategies that
+// arent Redis-list-based are added.
+type routingStrategy interface {
+
+	// Returns true if this routing strategy applies to the given routing
+	// parameters, false otherwise. Note: applies() must be deterministic.
+	applies(params routingParams) bool
+
+	// Returns the routing info (preferredNodeLimit and routingKey) for the
+	// provided routing parameters. The preferredNodeLimit is the number of
+	// preferred executor nodes that should be used, and the routing Key is the
+	// Redis key where the list of preferred executor nodes are stored.
+	routingInfo(params routingParams) (int, string, error)
+}
+
+// The runnerRecyclingRoutingStrategy is a routing strategy that attempts to
+// "recycle" warm execution nodes when possible.
+type runnerRecyclingRoutingStrategy struct {
```

Comment:
runnerRecyclingRoutingStrategy is long -- maybe just runnerRecycling? Similarly for OutputAffinity{}.

Reply:
SG, changed to `runnerRecycler` and `affinityRouter`